### ISSUE TITLE
[feature] Support using `-Wl,-hidden-l` instead of `-l` on macOS

### ISSUE THREAD

**Issue by mmomtchev** [conan-io/conan#15860]
### What is your suggestion?

Conan dependencies are usually built as static libraries - because they are meant to be used for a self-contained build.

When this build is producing a shared library - for example a Python or a Node.js module - or any other shared library - the symbols from these static libraries coming from conan are at risk of clashing with symbols from the loading application.

This problem is specific to POSIX systems, since Windows uses an entirely different system with `dllexport`/`dllimport`.

On Linux, the GNU linker supports linking with `-Wl,--exclude-libs,ALL` to prevent this issue.

On macOS, the LLVM linker also supports this, but requires to link individually each library with `-Wl,-hidden-l` instead of simply `-l`. This means that this must be done at the conan level, as each library option must be prefixed with this argument.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @mmomtchev 

Thanks for your suggestion

> On macOS, the LLVM linker also supports this, but requires to link individually each library with -Wl,-hidden-l instead of simply -l. This means that this must be done at the conan level, as each library option must be prefixed with this argument.

I am afraid these flags are compiler flags that would be defined by the build system, Conan is not defining them, so it would be very challenging for Conan to support this.
It is the responsibility of the build system scripts to do what they intend to do for shared libraries, and do the right thing (as every other build aspect), Conan is just wrapping and calling the build system like ``cmake ...``. Is there any way that this can be instructed to build system from the outside? It would be necessary to pass it to cmake, meson and autotools, possibly bazel too?

Because conan has the conf ``tools.build:cxxflags`` and linkflags too, but this doesn't work for linking with the libraries ``-l``, because this is decided by the build system based on the ``find_package()`` and ``target_link_libraries()`` or equivalent functionalities.

Am I missing something? Thanks!

**Comment by mmomtchev** [comment]
Yes, this is indeed the case when using `CMake` - I am using `pkg-config`.

Alas, fixing the problem would require some kind of API on `conan`'s side that all recipes would have to be modified to use to pass the name of the library so that `conan` can generate the appropriate link argument. All this for the case of using `pkg-config` on macOS when building a shared library.

`CMake` can indeed solve this in the consumer project's build.

**Comment by mmomtchev** [comment]
In fact, this can be solved in the `pkg-config` generator, I will give it a try
